waiter/src/waiter/service_description.clj

```diff
@@ -874,3 +888,25 @@
           (and (= "token" consent-mode) (= consent-id token) (= consent-owner owner)))
       (> (+ auth-timestamp (-> consent-expiry-days t/days t/in-millis))
          (.getMillis ^DateTime (clock))))))
+
+(defn service-id->idle-timeout
+  "Computes the idle timeout, in minutes, for a given service.
+   If the service is active or was created by on-the-fly, the idle timeout is retrieved from the service description.
+   Else, the idle timeout is the sum of the fallback period seconds and the outdated service timeout."
+  [service-id->service-description-fn token->token-hash token->token-data-field default-fallback-period-secs
+   outdated-service-timeout-mins service-id]
+  (let [{:strs [idle-timeout-mins source-tokens]} (service-id->service-description-fn service-id)]
+    (if (and (-> source-tokens count pos?)
+             (some (fn [{:strs [token version]}]
+                     (not= (token->token-hash token) version))
+                   source-tokens))
+      (do
+        (log/info service-id "that uses" (mapv #(get % "token") source-tokens) "is outdated")
+        (->> source-tokens
+             (map #(token->token-data-field (get % "token") "fallback-period-secs")) ;; get latest value
```

Comment:
We're loading the token data again for all the source tokens.  We did this previously in the `if`.  We should avoid assuming things are cached and load the data once, perhaps creating a `token->token-data` map that includes each of the `source-tokens` at the top of this function.